### PR TITLE
Avx

### PR DESCRIPTION
**Pull request by gwenn**
I had trouble implementing `_mm256_dp_ps`, `_mm_cmp_pd`, `_mm_cmp_sd` (I will try again).
I submit this PR because I don't want to prevent you from implementing other avx intrinsics but I will probably continue to commit in this branch until it is merged.

### CODE REVIEW

**Comment by gwenn** [comment]
I don't know why but `_mm256_zeroupper` and `_mm256_zeroall` generate "too many instructions in the disassembly" on windows platform:
https://ci.appveyor.com/project/rust-lang-libs/stdsimd/build/1.0.210
What should I do?

**Comment by alexcrichton** [comment]
Thanks for the PR! The error is indeed curious here. I wonder if the ABI for calling functions wrt SIMD registers is slightly different on Windows? In any case it looks like the right instructions are being generated, so want to up the instruction limit to maybe 30 instructions?

File: src/x86/avx.rs [hunk 1]
```diff
@@ -68,6 +70,71 @@ pub unsafe fn _mm256_or_ps(a: f32x8, b: f32x8) -> f32x8 {
     mem::transmute(a | b)
 }
 
+/// Shuffle double-precision (64-bit) floating-point elements within 128-bit
+/// lanes using the control in `imm8`.
+#[inline(always)]
+#[target_feature = "+avx"]
+//#[cfg_attr(test, assert_instr(vshufpd, imm8 = 0x0))] // FIXME
```
Comment (alexcrichton) [inline src/x86/avx.rs]: Should this be enabled now?

File: src/x86/avx.rs [hunk 2]
```diff
@@ -269,6 +336,383 @@ pub unsafe fn _mm256_sqrt_pd(a: f64x4) -> f64x4 {
...
+    match imm8 & 1 {
+        0 => simd_shuffle2(a, _mm256_undefined_si256(), [0, 1]),
+        _ => simd_shuffle2(a, _mm256_undefined_si256(), [2, 3]),
+    }
+}
+
+/// Extract an 8-bit integer from `a`, selected with `imm8`.
+#[inline(always)]
+#[target_feature = "+avx"]
+pub unsafe fn _mm256_extract_epi8(a: i8x32, imm8: i32) -> i32 {
```
Comment (alexcrichton) [inline src/x86/avx.rs]: Should this and some of the extractions below assert for a particular instruction?

**Comment by alexcrichton** [comment]
I also closed #91 and #95 as they accidentally duplicated the work from here, but maybe the tests could be included?

File: src/x86/avx.rs [hunk 3]
```diff
@@ -269,6 +336,383 @@ pub unsafe fn _mm256_sqrt_pd(a: f64x4) -> f64x4 {
...
+#[cfg_attr(test, assert_instr(vdpps, imm8 = 0x0))]
+pub unsafe fn _mm256_dp_ps(a: f32x8, b: f32x8, imm8: i32) -> f32x8 {
+    macro_rules! call {
+        ($imm8:expr) => { vdpps(a, b, $imm8) }
+    }
+    constify_imm8!(imm8, call)
+}
+
+/// Horizontally add adjacent pairs of double-precision (64-bit) floating-point
+/// elements in `a` and `b`, and pack the results.
```
Comment (vbarrielle) [inline src/x86/avx.rs]: Is it possible to include documentation for the location of the results, as in #95 ? Eg:

```rust
/// In the result, sums of elements from `a` are returned in even locations,
/// while sums of elements from `b` are returned in odd locations.
```

File: src/x86/avx.rs [hunk 4]
```diff
@@ -269,6 +336,383 @@ pub unsafe fn _mm256_sqrt_pd(a: f64x4) -> f64x4 {
...
+#[inline(always)]
+#[target_feature = "+avx"]
+#[cfg_attr(test, assert_instr(vhaddpd))]
+pub unsafe fn _mm256_hadd_pd(a: f64x4, b: f64x4) -> f64x4 {
+    vhaddpd(a, b)
+}
+
+/// Horizontally add adjacent pairs of single-precision (32-bit) floating-point
+/// elements in `a` and `b`, and pack the results.
```
Comment (vbarrielle) [inline src/x86/avx.rs]: Similarly, you could add

```rust
/// In the result, sums of elements from `a` are returned in locations of
/// indices 0, 1, 4, 5; while sums of elements from `b` are in locations
/// 2, 3, 6, 7.
```

File: src/x86/avx.rs [hunk 5]
```diff
@@ -269,6 +336,383 @@ pub unsafe fn _mm256_sqrt_pd(a: f64x4) -> f64x4 {
...
+#[inline(always)]
+#[target_feature = "+avx"]
+#[cfg_attr(test, assert_instr(vhsubpd))]
+pub unsafe fn _mm256_hsub_pd(a: f64x4, b: f64x4) -> f64x4 {
+    vhsubpd(a, b)
+}
+
+/// Horizontally subtract adjacent pairs of single-precision (32-bit) floating-point
+/// elements in `a` and `b`, and pack the results.
```
Comment (vbarrielle) [inline src/x86/avx.rs]:
```rust
/// In the result, sums of elements from `a` are returned in locations of
/// indices 0, 1, 4, 5; while sums of elements from `b` are in locations
/// 2, 3, 6, 7.
```

File: src/x86/avx.rs [hunk 6]
```diff
@@ -269,6 +336,383 @@ pub unsafe fn _mm256_sqrt_pd(a: f64x4) -> f64x4 {
...
+#[inline(always)]
+#[target_feature = "+avx"]
+#[cfg_attr(test, assert_instr(vhaddps))]
+pub unsafe fn _mm256_hadd_ps(a: f32x8, b: f32x8) -> f32x8 {
+    vhaddps(a, b)
+}
+
+/// Horizontally subtract adjacent pairs of double-precision (64-bit) floating-point
+/// elements in `a` and `b`, and pack the results.
```
Comment (vbarrielle) [inline src/x86/avx.rs]:
```rust
/// In the result, sums of elements from `a` are returned in even locations,
/// while sums of elements from `b` are returned in odd locations.
```

File: src/x86/avx.rs (hunk 1 above)
Comment (gwenn) [reply src/x86/avx.rs]: The [doc](https://gist.github.com/anonymous/25d752fda8521d29699a826b980218fc#file-intel-intrinsics-3-3-15-xml-L14103) says that `vshufpd` should be generated but [clang](https://github.com/llvm-mirror/clang/blob/master/test/CodeGen/avx-builtins.c#L1113-L1117) does not assert that it is the case. Maybe the expanded code is wrong ?
```rust
           match (imm8 >> 0) & 1 {
                0 =>
                match (imm8 >> 1) & 1 {
                    0 =>
                    match (imm8 >> 2) & 1 {
                        0 =>
                        match (imm8 >> 3) & 1 {
                            0 => simd_shuffle4(a, b, [0, 4, 2, 6]),
                            _ => simd_shuffle4(a, b, [0, 4, 2, 7]),
                        },
                        _ =>
                        match (imm8 >> 3) & 1 {
                            0 => simd_shuffle4(a, b, [0, 4, 3, 6]),
                            _ => simd_shuffle4(a, b, [0, 4, 3, 7]),
                        },
                    },
                    _ =>
                    match (imm8 >> 2) & 1 {
                        0 =>
                        match (imm8 >> 3) & 1 {
                            0 => simd_shuffle4(a, b, [0, 5, 2, 6]),
                            _ => simd_shuffle4(a, b, [0, 5, 2, 7]),
                        },
                        _ =>
                        match (imm8 >> 3) & 1 {
                            0 => simd_shuffle4(a, b, [0, 5, 3, 6]),
                            _ => simd_shuffle4(a, b, [0, 5, 3, 7]),
                        },
                    },
                },
                _ =>
                match (imm8 >> 1) & 1 {
                    0 =>
                    match (imm8 >> 2) & 1 {
                        0 =>
                        match (imm8 >> 3) & 1 {
                            0 => simd_shuffle4(a, b, [1, 4, 2, 6]),
                            _ => simd_shuffle4(a, b, [1, 4, 2, 7]),
                        },
                        _ =>
                        match (imm8 >> 3) & 1 {
                            0 => simd_shuffle4(a, b, [1, 4, 3, 6]),
                            _ => simd_shuffle4(a, b, [1, 4, 3, 7]),
                        },
                    },
                    _ =>
                    match (imm8 >> 2) & 1 {
                        0 =>
                        match (imm8 >> 3) & 1 {
                            0 => simd_shuffle4(a, b, [1, 5, 2, 6]),
                            _ => simd_shuffle4(a, b, [1, 5, 2, 7]),
                        },
                        _ =>
                        match (imm8 >> 3) & 1 {
                            0 => simd_shuffle4(a, b, [1, 5, 3, 6]),
                            _ => simd_shuffle4(a, b, [1, 5, 3, 7]),
                        },
                    },
                },
            }
```

File: src/x86/avx.rs (hunk 2 above)
Comment (gwenn) [reply src/x86/avx.rs]: It appears that there is [no instrunction specified](https://gist.github.com/anonymous/25d752fda8521d29699a826b980218fc#file-intel-intrinsics-3-3-15-xml-L14777-L14788) for this intrinsic.

File: src/x86/avx.rs (hunk 3 above)
Comment (gwenn) [reply src/x86/avx.rs]: I will try to merge your changes in my branch.

**Comment by alexcrichton** [comment]
Thanks @gwenn!

**Pull request merged**